TDataFrame.hxx

```diff
@@ -22,828 +22,1042 @@
-      auto FilterPtr = std::unique_ptr<DFF>(new DFF(f, actualBl, *fDerivedPtr));
-      auto& FilterRef = *FilterPtr;
-      Book(std::move(FilterPtr));
-      return FilterRef;
+   TDataFrameInterface() = delete;
+   TDataFrameInterface(std::shared_ptr<Proxied> proxied) : fProxiedPtr(proxied) {}
+   TDataFrameInterface(const std::string &treeName, TDirectory *dirPtr, const BranchList &defaultBranches = {});
+   TDataFrameInterface(TTree &tree, const BranchList &defaultBranches = {});
+
+   template <typename F>
+   TDataFrameInterface<Details::TDataFrameFilter<F, Proxied>> Filter(F f, const BranchList &bl = {})
+   {
+      ROOT::Internal::CheckFilter(f);
+      const BranchList &defBl = fProxiedPtr->GetDataFrame().lock()->GetDefaultBranches();
+      const BranchList &actualBl = Internal::PickBranchList(f, bl, defBl);
+      using DFF_t = Details::TDataFrameFilter<F, Proxied>;
+      auto FilterPtr = std::make_shared<DFF_t> (f, actualBl, fProxiedPtr);
+      TDataFrameInterface<DFF_t> tdf_f(FilterPtr);
+      Book(FilterPtr);
+      return tdf_f;
    }
 
-   template<typename F>
-   auto AddBranch(const std::string& name, F expression, const BranchList& bl = {})
-   -> TDataFrameBranch<F, Derived>& {
-      const BranchList& defBl = fDerivedPtr->GetDataFrame().GetDefaultBranches();
-      const BranchList& actualBl = ::PickBranchList(expression, bl, defBl);
-      using DFB = TDataFrameBranch<F, Derived>;
-      auto BranchPtr = std::unique_ptr<DFB>(new DFB(name, expression, actualBl, *fDerivedPtr));
-      auto& BranchRef = *BranchPtr;
-      Book(std::move(BranchPtr));
-      return BranchRef;
+   template <typename F>
+   TDataFrameInterface<Details::TDataFrameBranch<F, Proxied>> AddBranch(const std::string &name, F expression,
+                                                                        const BranchList &bl = {})
+   {
+      const BranchList &defBl = fProxiedPtr->GetDataFrame().lock()->GetDefaultBranches();
+      const BranchList &actualBl = Internal::PickBranchList(expression, bl, defBl);
+      using DFB_t = Details::TDataFrameBranch<F, Proxied>;
+      auto BranchPtr = std::make_shared<DFB_t>(name, expression, actualBl, fProxiedPtr);
+      TDataFrameInterface<DFB_t> tdf_b(BranchPtr);
+      Book(BranchPtr);
+      return tdf_b;
    }
 
-   template<typename F>
-   void Foreach(F f, const BranchList& bl = {}) {
-      const BranchList& defBl = fDerivedPtr->GetDataFrame().GetDefaultBranches();
-      const BranchList& actualBl = ::PickBranchList(f, bl, defBl);
-      using ArgTypes = typename FunctionTraits<decltype(f)>::ArgTypes;
-      using RetType = typename FunctionTraits<decltype(f)>::RetType;
-      auto fWithSlot = ::AddSlotParameter<RetType>(f, ArgTypes());
-      using DFA = TDataFrameAction<decltype(fWithSlot), Derived>;
-      Book(std::unique_ptr<DFA>(new DFA(fWithSlot, actualBl, *fDerivedPtr)));
+   template <typename F>
+   void Foreach(F f, const BranchList &bl = {})
+   {
+      GetDataFrameChecked();
```

Comment:
on this topic, shouldn't `GetDataFrameChecked` return a `shared_ptr` after checking for its validity once?